File: packages/node-core/src/tunnel-profiles.ts
import { randomUUID } from "node:crypto";
import { existsSync } from "node:fs";
import Database from "better-sqlite3";
import type { ProxyTunnelConfig, TransportLayerConfig } from "./connections.js";
import { parseListIndex, resolveProxyProfileByIndex } from "./list-index.js";
import { dbPath as defaultDbPath } from "./paths.js";

export type TunnelProfile = TransportLayerConfig;

export interface TunnelProfileStoreOptions {
  path?: string;
}

function openDb(readonly = false, path = defaultDbPath()): Database.Database {
  return new Database(path, { readonly });
}

function tunnelProfilesTableExists(db: Database.Database): boolean {
  const row = db.prepare("SELECT 1 FROM sqlite_master WHERE type = 'table' AND name = ?").get("tunnel_profiles") as { "1": number } | undefined;
  return !!row;
}

export async function loadTunnelProfiles(options: TunnelProfileStoreOptions = {}): Promise<TunnelProfile[]> {
  const path = options.path ?? defaultDbPath();
  if (!existsSync(path)) return [];

  const db = openDb(true, path);
  try {
    if (!tunnelProfilesTableExists(db)) return [];
    const rows = db.prepare("SELECT config_json FROM tunnel_profiles ORDER BY rowid").all() as { config_json: string }[];
    return rows.map((row) => JSON.parse(row.config_json) as TunnelProfile);
  } finally {
    db.close();
  }
}

export async function findTunnelProfile(id: string, options: TunnelProfileStoreOptions = {}): Promise<TunnelProfile | undefined> {
  const trimmed = id.trim();
  if (!trimmed) return undefined;
  const profiles = await loadTunnelProfiles(options);
  return profiles.find((profile) => profile.id === trimmed);
}

export function isProxyTunnelProfile(profile: TunnelProfile): profile is { type: "proxy" } & ProxyTunnelConfig {
  return profile.type === "proxy";
}

export interface ProxyProfileRefArgs {
  proxy_profile_id?: string;
  proxy_profile_name?: string;
}

export interface InlineProxyArgs {
  proxy_enabled?: boolean;
  proxy_host?: string;
  proxy_port?: number;
  proxy_username?: string;
  proxy_password?: string;
}

export function hasProxyProfileRef(args: ProxyProfileRefArgs): boolean {
  return !!(args.proxy_profile_id?.trim() || args.proxy_profile_name?.trim());
}

export function hasInlineProxyParams(args: InlineProxyArgs): boolean {
  return (
    !!args.proxy_enabled ||
    !!args.proxy_host?.trim() ||
    args.proxy_port !== undefined ||
    !!args.proxy_username?.trim() ||
    !!args.proxy_password
  );
}

export function findProxyProfilesByName(profiles: TunnelProfile[], name: string): TunnelProfile[] {
  const lower = name.trim().toLowerCase();
  return profiles.filter((profile) => profile.type === "proxy" && (profile.name || "").toLowerCase() === lower);
}

export function findProxyProfile(profiles: TunnelProfile[], args: ProxyProfileRefArgs): TunnelProfile | undefined {
  if (args.proxy_profile_id?.trim()) {
    const trimmed = args.proxy_profile_id.trim();
    const match = profiles.find((profile) => profile.id === trimmed);
    if (match?.type === "proxy") return match;
    const listIndex = parseListIndex(trimmed);
    if (listIndex !== undefined) return resolveProxyProfileByIndex(profiles, listIndex);
    return undefined;
  }
  if (args.proxy_profile_name?.trim()) {
    const trimmed = args.proxy_profile_name.trim();
    const matches = findProxyProfilesByName(profiles, trimmed);
    if (matches.length === 1) return matches[0];
    const listIndex = parseListIndex(trimmed);
    if (listIndex !== undefined) return resolveProxyProfileByIndex(profiles, listIndex);
    return undefined;
  }
  return undefined;
}

export function proxyProfileSummary(profile: { type: "proxy" } & ProxyTunnelConfig): string {
  if (profile.type !== "proxy") return profile.name || profile.id;
  if (!profile.host) return profile.name || profile.id;
  return `${profile.proxy_type || "socks5"}://${profile.host}:${profile.port || 1080}`;
}

/**
 * Builds the reference stub stored on a connection for a shared proxy profile.
 * Credentials stay in the profile; the backend resolves the reference at connect time.
 */
export function proxyProfileReferenceLayer(profile: { type: "proxy" } & ProxyTunnelConfig, layerId: string): TransportLayerConfig {
  return {
    type: "proxy",
    id: layerId,
    name: profile.name || "",
    enabled: profile.enabled !== false,
    profile_id: profile.id,
    proxy_type: profile.proxy_type || "socks5",
    host: "",
    port: 1080,
    username: "",
    password: "",
  };
}

type LegacyProxyFields = {
  proxy_enabled?: boolean;
  proxy_type?: "socks5" | "http";
  proxy_host?: string;
  proxy_port?: number;
  proxy_username?: string;
  proxy_password?: string;
};

/**
 * Replace any existing proxy (inline or profile stub) with a saved-profile reference.
 * Non-proxy layers (e.g. SSH) are preserved. Legacy proxy_* fields are cleared so
 * they cannot reintroduce a stacked proxy via normalizeTransportLayers.
 * Does not persist — callers decide whether to save or use for one request.
 */
export function applyProxyProfileOverride<T extends { transport_layers?: TransportLayerConfig[] }>(
  config: T,
  profile: { type: "proxy" } & ProxyTunnelConfig,
  layerId: string = randomUUID(),
): T {
  const existing = Array.isArray(config.transport_layers) ? config.transport_layers : [];
  const kept = existing.filter((layer) => layer.type !== "proxy");
  const next = {
    ...config,
    transport_layers: [...kept, proxyProfileReferenceLayer(profile, layerId)],
  } as T & LegacyProxyFields;
  next.proxy_enabled = false;
  delete next.proxy_type;
  delete next.proxy_host;
  delete next.proxy_port;
  delete next.proxy_username;
  delete next.proxy_password;
  return next;
}

File: crates/dbx-mcp/src/tunnel_profiles.rs
//! Saved tunnel / proxy profile helpers (DBX Settings → Tunnels).

use dbx_core::models::connection::{ProxyTunnelConfig, ProxyType, TransportLayerConfig};
use uuid::Uuid;

use crate::list_index::{item_at_list_index, parse_list_index};

pub type TunnelProfile = TransportLayerConfig;

#[derive(Debug, Clone, Default)]
pub struct ProxyProfileRefArgs {
    pub proxy_profile_id: Option<String>,
    pub proxy_profile_name: Option<String>,
}

#[derive(Debug, Clone, Default)]
pub struct InlineProxyArgs {
    pub proxy_enabled: Option<bool>,
    pub proxy_host: Option<String>,
    pub proxy_port: Option<u16>,
    pub proxy_username: Option<String>,
    pub proxy_password: Option<String>,
    pub proxy_type: Option<String>,
}

pub fn has_proxy_profile_ref(args: &ProxyProfileRefArgs) -> bool {
    args.proxy_profile_id.as_deref().is_some_and(|v| !v.trim().is_empty())
        || args.proxy_profile_name.as_deref().is_some_and(|v| !v.trim().is_empty())
}

pub fn has_inline_proxy_params(args: &InlineProxyArgs) -> bool {
    args.proxy_enabled == Some(true)
        || args.proxy_host.as_deref().is_some_and(|v| !v.trim().is_empty())
        || args.proxy_port.is_some()
        || args.proxy_username.as_deref().is_some_and(|v| !v.trim().is_empty())
        || args.proxy_password.as_ref().is_some_and(|v| !v.is_empty())
}

pub fn list_proxy_profiles(profiles: &[TunnelProfile]) -> Vec<&TunnelProfile> {
    profiles
        .iter()
        .filter(|profile| matches!(profile, TransportLayerConfig::Proxy(_)))
        .collect()
}

pub fn find_proxy_profiles_by_name<'a>(profiles: &'a [TunnelProfile], name: &str) -> Vec<&'a TunnelProfile> {
    let lower = name.trim().to_ascii_lowercase();
    list_proxy_profiles(profiles)
        .into_iter()
        .filter(|profile| profile.name().eq_ignore_ascii_case(&lower))
        .collect()
}

pub fn resolve_proxy_profile_by_index<'a>(profiles: &'a [TunnelProfile], index: usize) -> Option<&'a TunnelProfile> {
    let proxies = list_proxy_profiles(profiles);
    item_at_list_index(&proxies, index).copied()
}

pub fn find_proxy_profile<'a>(profiles: &'a [TunnelProfile], args: &ProxyProfileRefArgs) -> Option<&'a TunnelProfile> {
    if let Some(id) = args.proxy_profile_id.as_deref().map(str::trim).filter(|v| !v.is_empty()) {
        if let Some(match_by_id) = profiles.iter().find(|profile| {
            matches!(profile, TransportLayerConfig::Proxy(_)) && profile.id() == id
        }) {
            return Some(match_by_id);
        }
        if let Some(index) = parse_list_index(id) {
            return resolve_proxy_profile_by_index(profiles, index);
        }
        return None;
    }
    if let Some(name) = args.proxy_profile_name.as_deref().map(str::trim).filter(|v| !v.is_empty()) {
        let matches = find_proxy_profiles_by_name(profiles, name);
        if matches.len() == 1 {
            return Some(matches[0]);
        }
        if let Some(index) = parse_list_index(name) {
            return resolve_proxy_profile_by_index(profiles, index);
        }
    }
    None
}

pub fn proxy_profile_summary(profile: &TunnelProfile) -> String {
    match profile {
        TransportLayerConfig::Proxy(proxy) => {
            if proxy.host.is_empty() {
                if proxy.name.is_empty() {
                    proxy.id.clone()
                } else {
                    proxy.name.clone()
                }
            } else {
                let kind = match proxy.proxy_type {
                    ProxyType::Http => "http",
                    ProxyType::Socks5 => "socks5",
                };
                format!("{kind}://{}:{}", proxy.host, if proxy.port == 0 { 1080 } else { proxy.port })
            }
        }
        other => {
            if other.name().is_empty() {
                other.id().to_string()
            } else {
                other.name().to_string()
            }
        }
    }
}

pub fn proxy_profile_reference_layer(profile: &ProxyTunnelConfig, layer_id: String) -> TransportLayerConfig {
    TransportLayerConfig::Proxy(ProxyTunnelConfig {
        id: layer_id,
        name: profile.name.clone(),
        enabled: profile.enabled,
        proxy_type: profile.proxy_type,
        host: String::new(),
        port: 1080,
        username: String::new(),
        password: String::new(),
        profile_id: profile.id.clone(),
    })
}

pub fn apply_proxy_profile_override(
    mut config: dbx_core::models::connection::ConnectionConfig,
    profile: &ProxyTunnelConfig,
) -> dbx_core::models::connection::ConnectionConfig {
    let kept: Vec<_> = config
        .transport_layers
        .into_iter()
        .filter(|layer| !matches!(layer, TransportLayerConfig::Proxy(_)))
        .collect();
    let mut layers = kept;
    layers.push(proxy_profile_reference_layer(profile, Uuid::new_v4().to_string()));
    config.transport_layers = layers;
    config
}

pub fn inline_proxy_layer(args: &InlineProxyArgs) -> Result<TransportLayerConfig, String> {
    let host = args
        .proxy_host
        .as_deref()
        .map(str::trim)
        .filter(|v| !v.is_empty())
        .ok_or_else(|| "proxy_host is required when configuring an inline proxy.".to_string())?
        .to_string();
    let port = args.proxy_port.unwrap_or(1080);
    let proxy_type = match args.proxy_type.as_deref().unwrap_or("socks5").trim().to_ascii_lowercase().as_str() {
        "http" => ProxyType::Http,
        "socks5" | "socks" | "" => ProxyType::Socks5,
        other => return Err(format!("Unsupported proxy_type \"{other}\". Use socks5 or http.")),
    };
    Ok(TransportLayerConfig::Proxy(ProxyTunnelConfig {
        id: Uuid::new_v4().to_string(),
        name: format!("{host}:{port}"),
        enabled: args.proxy_enabled.unwrap_or(true),
        proxy_type,
        host,
        port,
        username: args.proxy_username.clone().unwrap_or_default(),
        password: args.proxy_password.clone().unwrap_or_default(),
        profile_id: String::new(),
    }))
}

pub fn format_proxy_list(profiles: &[TunnelProfile]) -> String {
    let proxies = list_proxy_profiles(profiles);
    if proxies.is_empty() {
        return "No proxy tunnel profiles configured in DBX Settings > Tunnels.".to_string();
    }
    let mut output =
        String::from("| # | ID | Name | Endpoint | Enabled |\n| --- | --- | --- | --- | --- |");
    for (idx, profile) in proxies.iter().enumerate() {
        let TransportLayerConfig::Proxy(proxy) = profile else {
            continue;
        };
        output.push_str(&format!(
            "\n| {} | {} | {} | {} | {} |",
            idx + 1,
            escape_cell(&proxy.id),
            escape_cell(&proxy.name),
            escape_cell(&proxy_profile_summary(profile)),
            if proxy.enabled { "yes" } else { "no" },
        ));
    }
    output
}

fn escape_cell(value: &str) -> String {
    value.replace('|', "\\|").replace('\n', " ")
}
